src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs

```diff
@@ -822,17 +822,121 @@ internal override TypeSymbol SetUnknownNullabilityForReferenceTypes()
                 }
             }
 
-            if (!haveChanges)
+            TypeSymbol result = this;
+            if (haveChanges)
             {
-                allTypeArguments.Free();
+                var definition = this.OriginalDefinition;
+                TypeMap substitution = new TypeMap(definition.GetAllTypeParameters(), allTypeArguments.ToImmutable());
+                result = substitution.SubstituteNamedType(definition);
+            }
+
+            allTypeArguments.Free();
+            return result;
+        }
+
+        internal override TypeSymbol MergeNullability(TypeSymbol other, VarianceKind variance, out bool hadNullabilityMismatch)
+        {
+            Debug.Assert(this.Equals(other, TypeCompareKind.IgnoreDynamicAndTupleNames));
+
+            if (!IsGenericType)
+            {
+                hadNullabilityMismatch = false;
                 return this;
             }
+
+            var allTypeParameters = ArrayBuilder<TypeParameterSymbol>.GetInstance();
+            var allTypeArguments = ArrayBuilder<TypeSymbolWithAnnotations>.GetInstance();
+            bool haveChanges = MergeTypeArgumentNullability(this, (NamedTypeSymbol)other, variance, allTypeParameters, allTypeArguments, out hadNullabilityMismatch);
+
+            NamedTypeSymbol result;
+            if (haveChanges)
+            {
+                TypeMap substitution = new TypeMap(allTypeParameters.ToImmutable(), allTypeArguments.ToImmutable());
+                result = substitution.SubstituteNamedType(this.OriginalDefinition);
+            }
             else
             {
-                TypeMap substitution = new TypeMap(this.OriginalDefinition.GetAllTypeParameters(),
-                                                   allTypeArguments.ToImmutableAndFree());
+                result = this;
+            }
+
+            allTypeArguments.Free();
+            allTypeParameters.Free();
+            return result;
+        }
 
-                return substitution.SubstituteNamedType(this.OriginalDefinition);
+        /// <summary>
+        /// Returns true if there were changes.
```

Comment:
>Returns [](start = 12, length = 7)

Please document what this method does (computes allTypeParameters and allTypeArguments for a merged constructed type, placing the result into caller-allocated array builders)